Remove section on `git revert` (#92)
ImperialCollegeLondon/intermediate_grad_school_git_course: _episodes/l1-04-rewriting_history.md

legitimate reasons why you might want to do that, from keeping the commit history clean
of unsuccessful attempts to do something to incorporate work done by someone else.

There are a number of reasons why you may need to change your commit history, for
example:

- You have already made a commit, but realise there are other changes you forgot to
  include
- You made a commit, but then changed your mind and want to remove this change from your
  history
- You want to "move" one or more commits so they are based on top of some other work
  (e.g. new changes made to the `main` branch)

This episode explores some of the commands `git` offers to manipulate the commit history
for your benefit and that of your collaborators. But, first, we will look at another
useful command.

### Set aside your work safely with `stash`

> It is not rare that, while you are working on some feature, you need to check something
> else in another branch. Very often this is the case when you want to try some
> contributor's code as part of a pull request review process (see next episodes). You
> can commit the work you are doing, but if it is not in a state ready to be committed,
> what would you do?

Reviewer: ```suggestion
It is not rare that, while you are working on some feature, you need to check something
else in another branch. Very often this is the case when you want to try some
contributor's code as part of a pull request review process (see next episodes). You
can commit the work you are doing, but if it is not in a state ready to be committed,
what would you do? Or you start working on a branch only to realise that it is not the one that you were planning to work on?
```

After you have made those last minute changes - and `staged` them, if needed - all you
need to do to amend the last commit while keeping the same commit message is:

```sh
git commit --amend --no-edit
```
{: .commands}

Or this:

```sh
git commit --amend -m "New commit message"
```
{: .commands}

if you want to write a new commit message.

Note that this will replace the previous commit with a new one -- the commit hash will
be different -- so this approach must not be used if the commit was already pushed to
the remote repository and shared with collaborators.

### Reset

The next level of complexity rewriting history is `reset`: it lets you redo the last (or
last few) commit(s) you made so you can incorporate more changes, fix an error you have
spotted and that is worth incorporating as part of that commit and not as a separate one
or just improve your commit message. Unlike `git revert`, `git reset` will

> retrospectively alter your commit history, so should **not** be used where you have

Reviewer: ```suggestion
retrospectively alter your commit history, so it should **not** be used when you have
```

erased**.

To undo just the last commit, you can do:

```sh
git reset --hard HEAD^
```
{: .commands}

Otherwise, to go back in time to a specific commit, you would do:

```sh
git reset --hard COMMIT_HASH
```
{: .commands}

> ## `reset` vs `revert`
>
> `git revert` was [discussed in the introductory course](https://imperialcollegelondon.github.io/introductory_grad_school_git_course/l1-02-committing-history/index.html).
>
> `reset` and `revert` both let you undo things done in the past, but they both have
> very different use cases.
>
> - `reset` uses brute force, potentially with destructive consequences, to
>   make those changes and is suitable only if the work has not been shared with others

> >   already. Use when you want to get rid of recent work you're not happy with and start

Reviewer: ```suggestion
>   already. Use it when you want to get rid of recent work you're not happy with and start
```

>   all over again.
> - `revert` is more lightweight and surgical, to target specific changes and creating

> >   new commits to history. Use when code has already been shared with others or when

Reviewer: ```suggestion
>   new commits to history. Use it when code has already been shared with others or when
```
